### ISSUE TITLE
AI: Upgrade TensorFlow from v1.15.2 to v2.18.0

### ISSUE THREAD

**Issue by lastzero** [photoprism/photoprism#222]
**As a Go developer, I want to upgrade to the latest version of the TensorFlow library, so that I get all the latest features and can use the latest hardware for faster inference.**

We are [currently using TensorFlow 1.15.2](https://dl.photoprism.app/tensorflow/linux/) and would like to upgrade to version 2.x (or later):

- https://github.com/galeone/tfgo
- https://github.com/tensorflow/tensorflow

Since the TensorFlow project team has decided to remove the Go library wrapper originally available in the [`tensorflow/tensorflow`](https://github.com/tensorflow/tensorflow) repository, the [`galeone/tfgo` project](https://github.com/galeone/tfgo) provides an alternative Go wrapper for TensorFlow v2. However, it may not support the very latest v2 release or needs to be upgraded first to support it (as it is an open-source project, you can submit a PR with the necessary changes).

**Related Issues:**

- https://github.com/photoprism/photoprism/issues/536
- https://github.com/photoprism/photoprism/issues/907
- https://github.com/photoprism/photoprism/issues/1090
- https://github.com/photoprism/photoprism/issues/4356
- https://github.com/photoprism/photoprism/issues/3438
- https://github.com/photoprism/photoprism/pull/2005

**Acceptance Criteria:**

- [ ] Identify all major steps necessary to upgrade our Docker image and the application code to TensorFlow 2 (e.g. check documentation of TensorFlow Go library)
- [ ] Note that we also support arm64, not just x86, so we need to be able to build TF2 for/on the Raspberry Pi 3/4 (64bit Linux)
- [ ] Test our existing models with TensorFlow 2 and make sure they continue to work as expected (run existing unit tests and do manual acceptance tests)
- [ ] Document your findings here in this ticket plus add a summary to our wiki (stream of issue comments can be confusing to use as todo list to perform the actual upgrade later)

*Our [Developer Guide](https://docs.photoprism.app/developer-guide/) explains how to [set up a local build environment](https://docs.photoprism.app/developer-guide/setup/). You are welcome to [contact us](https://www.photoprism.app/contact) via email or chat if you would like to work on this and need additional guidance to get started.*

**Comment by guysoft** [comment]
I can help with the Rpi build.
You could also use qemu to build, but it might take days, so I would recommend just using a Pi.

**Comment by lastzero** [comment]
Good to know qemu is so slow! So not an option for us in practice.

We have a Raspberry Pi 4 with 4 GB RAM and an Nvidia Jetson in our lab for development.

**Comment by guysoft** [comment]
> Good to know qemu is so slow! So not an option for us in practice.
> 
Its usually OK, but tensorflow took 7 hours in Raspberrypi 4 running on two threads. If you have a super strong intel machine it might be worth it, but I'd recommend using a Pi just for now, since its a once-per-version build.

**Comment by lastzero** [comment]
We're also building the `photoprism-arm64` image on the Raspberry Pi 4 using your 64bit Ubuntu image at the moment. It's reasonable fast, so the question is if it's worth finding a different (cloud based) solution.

**Comment by guysoft** [comment]
You could also use multi-arch to build it using docker hub. If you use a local x86_64 machine it might also build.

**Comment by savegame** [comment]
i built your project in FreeNAS Jail (FreeBSD 11.2 at this moment) , but it has only Tensorflow 1.14 :) it would be sad if i cant build tensorflow 2 in freebsd ... is it possible make it compatible with tensorflow v1 and v2 in same time?

**Comment by lastzero** [comment]
Where do you have TF 1.14 from? We'll probably stick with it until after our first stable release.

**Comment by savegame** [comment]
i build TF 1.14 from FreeBSD ports

**Comment by lastzero** [comment]
> i build TF 1.14 from FreeBSD ports

Can you check for us if / when they have TF2? I've used FreeBSD many years ago. You're also welcome to write a howto for our docs.

**Comment by savegame** [comment]
i'll try later write same howto about FreeBSD (especially FreeNAS jail) (dont have much time :disappointed:  )

**Comment by McQueen2063** [comment]
@savegame Sorry for being a bit off topic in this thread, but I try to build photoprism under FreeBSD as well (that is, a FreeBSD 12.1 jail, not inside FreeNAS, but this shouldn't matter).
I did install tensorflow from pkg (py37-tensorflow-1.14.0_3), but `gmake all` later fails with:
```
go build -v ./...
github.com/tensorflow/tensorflow/tensorflow/go
# github.com/tensorflow/tensorflow/tensorflow/go
ld: error: unable to find library -ltensorflow
clang: error: linker command failed with exit code 1 (use -v to see invocation)
gmake: *** [Makefile:59: dep-go] Error 2
```

Well, yes, there is no libtensorflow.so.
How did you proceed from there?

**Comment by lastzero** [comment]
You might need to build it from source. Takes a couple of hours to compile though.

**Comment by McQueen2063** [comment]
I'm building `/usr/ports/science/py-tensorflow` right now. I wonder why the outcome of building the port is different than the pkg, but I'll give it a try anyways :)

**Comment by lastzero** [comment]
You need TensorFlow for C, not Python.

**Comment by lastzero** [comment]
As of right now, TensorFlow for Go seems broken for the average developer:

https://github.com/tensorflow/tensorflow/issues/41808

Their downloadable C library is also outdated. CPU requirements not really documented as well.

**Comment by guysoft** [comment]
Note there is a workaround: https://github.com/tensorflow/tensorflow/issues/41808#issuecomment-683362089
And it does work in master: https://github.com/tensorflow/tensorflow/issues/41808#issuecomment-696397820

**Comment by lastzero** [comment]
Yeah, but we really don't have time for hacks and workarounds right now. Need to compile it for different CPU architectures as well, including GPU support. Would be OK doing this as a full time job if there was nothing else to do.

**Comment by gyto6** [comment]
In case you need it, I think this link might help to prepare TF2 (maybe 3) support for Nvidia GPU.
https://pixinsight.com/forum/index.php?threads/gpu-accelerated-starnet-v2-working-with-cuda-and-libtensorflow-gpu-under-linux.18180/

**Comment by lastzero** [comment]
Since Google / the TensorFlow team decided to make it harder to use TensorFlow with Go (wonder why...), the next best/only option is to migrate our backend to this fork maintained by @galeone:

- https://github.com/galeone/tensorflow

**Ideally, we do this when the fork is getting updated to the then latest version in the upcoming weeks/months. Waiting is fine if it saves us from having to upgrade to a new TF version twice in short order.**

In our particular situation, upgrading also means that we (may) have to build and distribute new C libraries for all supported architectures and operating systems. This can be a lot of extra work, unless there are existing binaries, which has often not been the case in the past. We therefore plan accordingly, as hoping for luck is not a strategy.

**Comment by galeone** [comment]
Eventually, I updated the fork today.

You can depend on https://github.com/galeone/tensorflow/tree/r2.9-go

**Comment by lastzero** [comment]
Wow, that was quick! Thanks a lot! So now it's up to us to update our dependencies and refactor existing code as needed 🥳 

Do you happen to know if binary models exported with TF 1.x remain compatible or how to convert them if needed?

**Comment by galeone** [comment]
In TensorFlow 2.x you can only use the SavedModel serialization format. Hence, if you have some "frozen" models you can't import and use them.

**Comment by galeone** [comment]
FYI I updated https://github.com/galeone/tensorflow to TensorFlow 2.10 (https://github.com/galeone/tensorflow/tree/r2.10-go)

You can update via `go get -u github.com/galeone/tensorflow/tensorflow/go@r2.10-go`

Moreover, if you use (or plan to use) [galeone/tfgo](https://github.com/galeone/tfgo) the dependency of tfgo on galeone/tensorflow is already updated.

**Comment by lastzero** [comment]
@galeone Thanks for the update! I really appreciate it! Due to limited resources, we unfortunately have to wait with the upgrade until we have released the extended multi-user support.

**Comment by lapo-luchini** [comment]
> Can you check for us if / when they have TF2? I've used FreeBSD many years ago. You're also welcome to write a howto for our docs.

FreeBSD Ports seems to have [tensorflow 2.9.1](https://www.freshports.org/science/py-tensorflow/) but tensorflox 1.x never made in officially (though I have a personal binary pkg for that [for personal use published on GitHub](https://github.com/lapo-luchini/photoprism-freebsd-port/releases/tag/2023-07-19)) because [porting bazel-based stuff seems to be quite complex](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=266130#c7) and in the end no one finished it properly for an old version, so I look forward to when 2.x will be supported too. 😁

**Comment by lastzero** [comment]
@lapo-luchini Thank you for letting us know! We appreciate any help we can get. However, I understand that migrating to TF2 might be too much of a challenge for contributors. I'll otherwise give it a try once we've finished adding support for OAuth2, OIDC, and 2FA (see e.g. https://github.com/photoprism/photoprism/issues/3943) and have resolved the known issues related to facial recognition.

**Comment by lastzero** [comment]
Since I unfortunately still haven't had time to work on this issue, I moved it to [Help Wanted](https://github.com/orgs/photoprism/projects/5) and updated the description to provide additional context. If you would like to contribute and need help, please feel free to [contact us](https://www.photoprism.app/contact) via email or chat.

**Comment by raystlin** [comment]
Somewhere inside tensorflow docs they mention the existence of a community maintained project that compiles the go bindings of the tensorflow c API: https://github.com/wamuir/graft
I downloaded nasnet model, updated the Dockerfile at docker/tensorflow to compile tensorflow 2.17.0 on x64 (regular one, I did not try gpu support) and took the minimal amount of go code from your repository (internal/ai/classify) to compile a PoC. Without changing anything from your base code it seems to work.
My next attempt will be to cross compile it to arm64 and see if the same simple test works on a rbpi. If so, I will let you know.

**Comment by raystlin** [comment]
**tldr**: It works on arm64 as well (TF version 2.17.0 and graft v0.9.1), so I will try now to change photoprism sources, compile everything and pass the tests.
I would like to explain the process I followed to compile the library and the test. Bazel is supposed to simplify tasks like cross compilation, so I decided to inspect the file `.bazelrc` on tensorflow project to see if I could find any configuration to cross compile, and I found something:
```
# START LINUX AARCH64 CROSS-COMPILE CONFIGS
build:cross_compile_linux_arm64 --config=cross_compile_base

# Set the target CPU to Aarch64
build:cross_compile_linux_arm64 --platforms=//tensorflow/tools/toolchains/cross_compile/config:linux_aarch64
build:cross_compile_linux_arm64 --cpu=aarch64
build:cross_compile_linux_arm64 --crosstool_top=//tensorflow/tools/toolchains/cross_compile/cc:cross_compile_toolchain_suite
```
So, I tried to execute that configuration with
```
bazel build -c opt --config cross_compile_linux_arm64 --config monolithic //tensorflow:libtensorflow.so
```
And it did not work due to missing basic header files. Obviously the compilation environment was not the one it was expecting. I took a second look and I saw that there is a parent configuration named `cross_compile_base`, that we can find also on the same file.
```
build:cross_compile_base --host_cpu=k8
build:cross_compile_base --host_crosstool_top=//tensorflow/tools/toolchains/cross_compile/cc:cross_compile_toolchain_suite
build:cross_compile_base --extra_execution_platforms=//tensorflow/tools/toolchains/cross_compile/config:linux_x86_64
```
Which sets `host_cpu` to `k8`. Also `tensorflow/tools/toolchains/cross_compile/cc/BUILD` tells us it is using `aarch64_unknown_linux_gnu` toolchain, which is not the default one on debian, and `/dt10/` as an include directory. So it seems they are using a docker container an kubernetes to perform cross compilations.
The file `tensorflow/tools/toolchains/cross_compile/config/BUILD` points to `docker://gcr.io/tensorflow-testing/ml-devinfra-linux-aarch64-cross-compile@sha256:11c5ac3b9b4e01cfa82b39b90826a9bfc5b806ccc92cd3d272e6bf861de43be1` which is pullable and executable on x64 (though it is a 6GB image). On that image the former bazel build command works and cross compiles a valid `libtensorflow.so` library.
The final step was to cross compile the go code to arm64 and that is done using `GOARCH` environment variable... however, and for some reason I still don't know, it does not work. On compiling the test it does not recognize the basic types defined on graft.
I decided to move to an arm64 virtual machine using `qemu`, and `armbian` project works like a charm:
```
git clone https://github.com/armbian/build
cd build
./compile.sh BOARD=qemu-uboot-arm64 BRANCH=current RELEASE=bookworm
```
The exact same test compiled on a native arm64 environment it finally worked.

**Comment by raystlin** [comment]
I have run the tests and everything seems to be fine. However I have some doubts I would like to ask before doing a pull request.
# Regarding the docker images
1. I really don't know how you use them, so I am guessing what they should look like. I removed the download of the source code from `Dockerfile` as there exists make download.
2. `Jetson` and `arm64` seem to be more or less the same except for the names. Maybe it could be an environment variable passed to docker run. 
3. For `osx` I guess everything must change now that it has arm64 and amd64 flavors. I kept it as it was (adapting the `Makefile`) but maybe it should change.
4. Maybe I could unify the `Dockerfile` and `Makefile` definitions in one and detect on runtime the architecture to either make the `arm64` or the `amd64` builds.
5. I added a cross folder to cross compile `arm64` on `x64`, as I explained on my previous comment.

# Regarding  optimizations
1. I kept the original optimizations for `avx` and `avx2` on `x64`. I read `gcc` docs and it seems `sse4.2` and `sse4.1` get included on adding `avx2`... `fma` I don't think is included but I kept everything as it was as it seemed something balanced between number of versions and fine-grain optimizations.
2. I did not see a `cuda` enabled version on the `Makefile` options. I guess it was not available. Maybe I should include it?
3. Intel optimizations are documented by `tensoflow` docs and it seems everything should work if we add `--config mkl`. However, I tried and compilation failed due to a type conversion error on `xla`. I guess I could compile without `xla` but I don't know if it is worth it. I checked the release notes for `tensorflow` and on 2.15.0 it says:
```
[oneDNN CPU performance optimizations](https://github.com/tensorflow/community/blob/master/rfcs/20210930-enable-onednn-ops.md) Windows x64 & x86.

    Windows x64 & x86 packages:
        oneDNN optimizations are enabled by default on X86 CPUs
    To explicitly enable or disable oneDNN optimizations, set the environment variable TF_ENABLE_ONEDNN_OPTS to 1 (enable) or 0 (disable) before running TensorFlow. To fall back to default settings, unset the environment variable.
    oneDNN optimizations can yield slightly different numerical results compared to when oneDNN optimizations are disabled due to floating-point round-off errors from
    different computation approaches and orders.
    To verify if oneDNN optimizations are on, look for a message with "oneDNN custom operations are on" in the log. If the exact phrase is not there, it means they are off.
```
And on the issues I read `mkl` is no longer supported, so I guess that to get this optimization the only thing I can do is to compile an older version.
Also I saw that this exists https://github.com/intel/intel-extension-for-tensorflow but it is also a bit outdated:

|Intel® Extension for TensorFlow*| Stock TensorFlow|
| ------------------------------------------------| ------------------------ |
|Not support 	                              | 2.17                        |
|latest build from source                 | 2.16                        |
|v2.15.0.0 & v2.15.0.1                       | 2.15                        |
|v2.14.0.1 & v2.14.0.2                       | 2.14                        |
|v2.13.0.0                                            | 2.13                        |
|v1.2.0                                                 | 2.12                        |
|v1.1.0 	                                              | 2.10 & 2.11            |
|v1.0.0                                                 | 2.10                        |

**Comment by lastzero** [comment]
@raystlin Thank you very much! I'll be happy to test this when we're back in the office next week. Have you pushed your changes to a public repo from where we can pull and merge them? 🤔

**Comment by raystlin** [comment]
Yes, I forked the repo and created the branch `feature/tensorflow-2.x`

https://github.com/raystlin/photoprism/tree/feature/tensorflow-2.x

**Comment by lastzero** [comment]
@raystlin Awesome, thanks for working on this! I haven't found any roadblocks yet, so we'd be happy to merge your changes if you open a pull request for it 👍

**Comment by lastzero** [comment]
I found that Google (still) provides C library binaries for download at https://www.tensorflow.org/install/lang_c. However, there is (still) no Linux build for ARM64 (and ARMv7, if supported by TF2?).

So it looks like we would have to build our own ARM binaries, but could otherwise use the binaries that tensorflow.org provides? ⏳

**Comment by lastzero** [comment]
I've started TensorFlow 2.18.0 builds for amd64 and arm64. Let's see how long it takes! The last arm64 build took several days on the (low powered) Nvidia Jetson 😴

![Image](https://github.com/user-attachments/assets/29cda53d-79c9-4e1f-a830-44ef990c75d9)

**Comment by lastzero** [comment]
I'm happy to report that the TensorFlow v2.18.0 builds were successful so the binaries are now available for download:

- https://dl.photoprism.app/tensorflow/amd64/
- https://dl.photoprism.app/tensorflow/arm64/

Note that the memory requirements for this were quite high, so I had to add 16GB of swap to the 16GB of physical memory our arm64 build server has. Also, we don't have an `armv7-a` version yet, as the arm64 build environment only supports the following architectures: `armv8-a armv8.1-a armv8.2-a armv8.3-a armv8.4-a armv8.5-a armv8.6-a armv8.7-a armv8.8-a armv8.9-a armv8-r armv9-a armv9.1-a armv9.2-a armv9.3-a armv9.4-a`

**Comment by lastzero** [comment]
You can now test this with the development preview builds we provide:

- https://docs.photoprism.app/release-notes/#development-preview
- https://docs.photoprism.app/getting-started/updates/#development-preview

**Comment by lastzero** [comment]
There is one remaining issue with the `create_archive.sh` scripts used to package the binaries in `/docker/tesorflow`, as the .tar.gz archives contain copies of the same library binaries instead of symlinks (which multiplies the size of the archive and the size they consume when installed in production Docker images). Also, it includes `*.params` text files, that don't seem to be needed or helpful.

The contents of the lib/ directory in the `tar.gz` file **currently** look as follows:

```
.r-xr-xr-x 468M michael  2 Apr 09:44 libtensorflow.so
.r-xr-xr-x 468M michael  2 Apr 09:44 libtensorflow.so.2
.r-xr-xr-x 468M michael  2 Apr 09:44 libtensorflow.so.2.18.0
.r-xr-xr-x 875k michael  2 Apr 09:44 libtensorflow.so.2.18.0-2.params
.r-xr-xr-x  49M michael  2 Apr 09:44 libtensorflow_framework.so.2
.r-xr-xr-x  49M michael  2 Apr 09:44 libtensorflow_framework.so.2.18.0
.r-xr-xr-x 240k michael  2 Apr 09:44 libtensorflow_framework.so.2.18.0-2.params
```

However, this is what it **should** be (I manually fixed it for the packages available on dl.photoprism.app):

```
lrwxrwxrwx    - michael  2 Apr 11:02 libtensorflow.so -> libtensorflow.so.2.18.0
lrwxrwxrwx    - michael  2 Apr 11:02 libtensorflow.so.2 -> libtensorflow.so.2.18.0
.r-xr-xr-x 468M michael  2 Apr 09:44 libtensorflow.so.2.18.0
lrwxrwxrwx    - michael  2 Apr 11:02 libtensorflow_framework.so -> libtensorflow_framework.so.2.18.0
lrwxrwxrwx    - michael  2 Apr 11:02 libtensorflow_framework.so.2 -> libtensorflow_framework.so.2.18.0
.r-xr-xr-x  49M michael  2 Apr 09:44 libtensorflow_framework.so.2.18.0
```

It would be good if we could fix this before building the next version, so it doesn't happen again. However, in the grand scheme of things, this is really just a minor issue. Overall, the approach looks and works great so far! 🌈💎💐

**Comment by raystlin** [comment]
You are so right, I completely missed that one. I will fix it on my fork and make a PR.

**Comment by lastzero** [comment]
@raystlin Thank you so much! I think it is easiest if you create a new fork from our develop branch and then work on that, since I've already applied many of your changes manually (so I could use the night to compile the binaries).

**Comment by lastzero** [comment]
@raystlin Sorry to bother you again! I have two more questions that you may be able to answer faster than I can figure things out:

1. While the Ubuntu 24.10 base images for the build environments are great in that they include the latest (stable) compilers and tools, they are based on glibc 2.40 and so may prevent the binaries from working with Ubuntu 24.04 LTS (which includes glibc 2.39) and *any* older distributions that use **`glibc < 2.40`**. That is, unless the glibc compatibility requirements have changed OR libtensorflow is not linked against glibc (which I haven't had time to check, but would surprise me)? If so, would it make sense to (also) build more compatible libs (at least the cpu-noavx variant) on an older distribution like Ubuntu 22.04 LTS (Jammy)?
 
2. Unfortunately, the arm64 build image did not allow me to build binaries for **`armv7-a`**, as this is not on the long list of supported arm architectures (including the latest armv9). Can support for armv7 be added by installing additional packages, or would it make sense to add a dedicated armv7 build image to run on our arm64 server? Which strategy would you recommend? I'd rather not cross compile it with qemu for performance reasons (i.e. lack of time).

**Comment by raystlin** [comment]
1. Yes, it makes a lot of sense as glibc is backwards compatible. Maybe I have to change some dependencies, but I think it should work fine on jammy. I will test it and add it to the PR.
2. I am going to try to cross compile it from x64 to armhf (without qemu, just the toolchain) and see if it works, as doesn't seem to be directly supported by the configurations in .bazelrc. I will let you know the results.

**Comment by raystlin** [comment]
I am currently working on the armv7 compilation. The problem is the toolchain for armv7 got deprecated on tf 2.2.0 (more or less) and now armv7 is only supported for iOS and tflite on embedded devices. I am still investigating to cross compile from an arm64 host (as bazel deprecated arm support long ago and I have not yet investigated cake compilations for TF) but I only got errors so far...
I will make a new PR migrating arm64 docker images to jammy and adding (or not if I am not able to do it) a make target for armv7.

**Comment by lastzero** [comment]
If armv7 is no longer (properly) supported, then that's the way it is. We just need to communicate it so that affected users know they will not get updates anymore...

**Comment by raystlin** [comment]
I am afraid I reached a dead end. I managed to get a working arm toolchain but then the compiler hit this line of code on file `third_party/xla/xla/tsl/concurrency/async_value.h`:

```c
 struct WaitersAndState {
    // We rely on the fact that all `NotifierListNode` values are aligned at
    // least to 8 bytes and we can encode state in the lowest 3 bits. We use
    // the conservative estimation of the minimal alignment of pointers returned
    // from memory allocation functions.
    //
    // See: https://en.cppreference.com/w/cpp/types/max_align_t
    static_assert(alignof(std::max_align_t) >= 8 &&
                  sizeof(NotifierListNode*) == 8);
```

The pointer size in 32 bit arm (and intel) is 4. Furthermore, the comment hints they are using an optimization technique called `tagged pointers`, which uses the fact that 64 bit architectures align the memory to add state information using pointer bits that are always 0 (this is common in interpreters and JIT compilers).
I tried to compile without xla by setting `--define=with_xla_support=false` but it did not work.

I am afraid I ran out of ideas. Anyway, event if I could find a way to make it work, it would be hackish and could break at any moment... so I guess you are right and it is the way it is.

**Comment by raystlin** [comment]
That being said, I am going to make a PR to change the remaining Dockerfiles to jammy.

**Comment by lastzero** [comment]
Thanks for trying! We can create an armv7 branch that uses the old TensorFlow version if someone really needs it.

Note that I've just pushed a bunch of changes/improvements so that the archive names match the actual names at https://dl.photoprism.app/tensorflow/amd64/ and we can also offer a version for server CPUs with AVX512. The AMD64 docker images I created based on this have been [pushed to Docker Hub](https://hub.docker.com/repository/docker/photoprism/tensorflow/tags).

However, there is no ARM64 image on Docker Hub yet, and I haven't had time to look up/test any ARM64 optimizations. The compiler flags for AMD64 look reasonable, but I may have missed things because I didn't have time to do more research, e.g. I'm not sure if it makes sense to enable AVX2 and AVX512 instructions at the same time...

**Comment by lastzero** [comment]
@raystlin Do you think there should be a symbolic link from `libtensorflow_framework.so` to `libtensorflow_framework.so.2.18.0` in the binary archives, just like for `libtensorflow.so`? I'm not sure what it is for. Might [only be used by Python](https://github.com/tensorflow/tensorflow/issues/30488), which could be why we don't see an error? 🤔

**Comment by raystlin** [comment]
`libtensorflow_framework.so.2` is referenced by `libtensorflow.so` as 
```
 0x0000000000000001 (NEEDED)             Shared library: [libtensorflow_framework.so.2]
```
And graft directly imports `libtensorflow.so` in `tensorflow/lib.go`:
```go
package tensorflow

// #cgo LDFLAGS: -ltensorflow
import "C"
```

`libtensorflow_framework.so` symlink is created for programs that need to use the library directly. Graft is not and it does not seem it is going to do it, but keeping the symlink is almost free...

**Comment by lastzero** [comment]
Excellent! In this case, I won't add the link to the newly built binary archives I just uploaded:

- https://dl.photoprism.app/tensorflow/amd64/

All that's missing then are Jammy compatible ARM64 binaries, as we need them to build the installation packages for those who don't use our Docker images:

- https://dl.photoprism.app/pkg/linux/README.html

**Comment by raystlin** [comment]
I think I can have it tested and the PR done by the end of the day.

**Comment by lastzero** [comment]
@raystlin Thanks a lot! Your PR has been merged, and new ARM64 binaries should be available soon once the build is complete. 🥳 

I've meanwhile refactored our ai/model packages to make models configurable (instead of hardcoded):

- https://github.com/photoprism/photoprism/issues/127#issuecomment-2781700031

Unfortunately, it seems that image data needs to be transformed in different ways for use with TensorFlow, as some models e.g. expect pixel values to be floats in the range [-1, 1] and others [0, 1]). This means that you cannot simply swap or add models, so adding support for other models will require more work...

**Comment by lastzero** [comment]
The new ARM64 build is now available for download at [dl.photoprism.app/tensorflow/arm64/](https://dl.photoprism.app/tensorflow/arm64/). It's about 4 MB smaller than the last one, thanks to the newly added `--copt=-O2` compile flag (`--copt=-O3` makes the binaries much larger, so I think we'll stick with that, and based on their size, it seems the TensorFlow team uses it for their [official packages](https://www.tensorflow.org/install/lang_c) too). With that, I believe we can finally consider the upgrade complete...? 🥳 

**A big thank you to @raystlin for figuring out the details, the pull requests, and the advice! Your help was invaluable.** 🙏
